runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+			let bytes = env.ext_id().to_le_bytes();
+			(bytes[0], bytes[1])
+		};
+
+		let result = match FuncId::try_from(function_id) {
+			Ok(function_id) => {
+				// Read encoded parameters from buffer and calculate weight for reading `len` bytes`.
+				// reference: https://github.com/paritytech/polkadot-sdk/blob/117a9433dac88d5ac00c058c9b39c511d47749d2/substrate/frame/contracts/src/wasm/runtime.rs#L267
+				let len = env.in_len();
+				env.charge_weight(contract_host_weight.return_per_byte.saturating_mul(len.into()))?;
+				let params = env.read(len)?;
+				log::debug!(target: LOG_TARGET, "Read input successfully");
+				match function_id {
+					FuncId::Dispatch => {
+						dispatch::<T, E>(&mut env, version, pallet_index, call_index, params)
+					},
+					FuncId::ReadState => {
+						read_state::<T, E>(&mut env, version, pallet_index, call_index, params)
+					},
+					// TODO
+					FuncId::SendXcm => send_xcm::<T, E>(&mut env),
+				}
+			},
+			Err(e) => Err(e),
+		};
+
+		match result {
+			Ok(_) => Ok(RetVal::Converging(0)),
+			Err(e) => Ok(RetVal::Converging(convert_to_status_code(e, version))),
+		}
+	}
+}
+
+fn dispatch<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<(), DispatchError>
+where
+	T: frame_system::Config<RuntimeOrigin = RuntimeOrigin, RuntimeCall = RuntimeCall>,
+	RuntimeOrigin: From<RawOrigin<T::AccountId>>,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " dispatch |";
+	let call = construct_call(version, pallet_index, call_index, params)?;
+	// Contract is the origin by default.
+	let origin: RuntimeOrigin = RawOrigin::Signed(env.ext().address().clone()).into();
+	dispatch_call::<T, E>(env, call, origin, LOG_PREFIX)
+}
+
+fn dispatch_call<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	call: RuntimeCall,
+	mut origin: RuntimeOrigin,
+	log_prefix: &str,
+) -> Result<(), DispatchError>
+where
+	T: frame_system::Config<RuntimeOrigin = RuntimeOrigin, RuntimeCall = RuntimeCall>,
+	RuntimeOrigin: From<RawOrigin<T::AccountId>>,
+	E: Ext<T = T>,
+{
+	let charged_dispatch_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+	log::debug!(target:LOG_TARGET, "{} Inputted RuntimeCall: {:?}", log_prefix, call);
+	origin.add_filter(AllowedApiCalls::contains);
+	match call.dispatch(origin) {
+		Ok(info) => {
+			log::debug!(target:LOG_TARGET, "{} success, actual weight: {:?}", log_prefix, info.actual_weight);
+			// Refund weight if the actual weight is less than the charged weight.
+			if let Some(actual_weight) = info.actual_weight {
+				env.adjust_weight(charged_dispatch_weight, actual_weight);
+			}
+			Ok(())
+		},
+		Err(err) => {
+			log::debug!(target:LOG_TARGET, "{} failed: error: {:?}", log_prefix, err.error);
+			Err(err.error)
+		},
+	}
+}
+
+fn construct_call(
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<RuntimeCall, DispatchError> {
+	match pallet_index {
+		52 => {
```

Comment:
#93

Reply:
As this is in the runtime, this can probably be changed to super::Assets::index() to obtain the index of the pallet in the runtime in a strongly typed manner. This should be the preferred approach over constants within the runtime.